runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -0,0 +1,229 @@
+     */
+    @Getter
+    @Setter
+    Function<Object,Object> valueMutator = (x) -> x;
+
+    /** Batch size: number of SMREntry in a single CONTINUATION.
+     */
+    @Getter
+    @Setter
+    private int batchSize = 50;
+
+    /** BiConsumer to run after every CheckpointEntry is appended
+     * to the stream
+     */
+    @Getter
+    @Setter
+    BiConsumer<CheckpointEntry,Long> postAppendFunc = (cp, l) -> {};
+
+    /** Local ref to the object's runtime.
+     */
+    private CorfuRuntime rt;
+
+    /** Local ref to the stream's view.
+     */
+    StreamsView sv;
+
+    /** Local ref to the object that we're dumping.
+     *  TODO: generalize to all SMR objects.
+     */
+    private SMRMap map;
+
+    public CheckpointWriter(CorfuRuntime rt, UUID streamID, String author, SMRMap map) {
+        this.rt = rt;
+        this.streamID = streamID;
+        this.author = author;
+        this.map = map;
+        checkpointID = UUID.randomUUID();
+        // sv = new BackpointerStreamView(rt, streamID);
+        sv = rt.getStreamsView();
+    }
+
+    /** Static method for all steps necessary to append checkpoint
+     *  data for an SMRMap into its own stream.  An optimistic
+     *  read-only transaction is used to freeze the contents of
+     *  the map while the checkpoint entries are written to
+     *  the stream.
+     */
+
+    public List<Long> appendCheckpoint() {
+        return appendCheckpoint(cpw -> {});
+    }
+
+    /**
+     * @param setupWriter Lambda to transform the new CheckPointWriter
+     *                    object (e.g., change batch size, etc) prior
+     *                    to use.
+     * @return List of global addresses of all entries for this checkpoint.
+     */
+
+    public List<Long> appendCheckpoint(Consumer<CheckpointWriter> setupWriter) {
+        List<Long> addrs = new ArrayList<>();
+        setupWriter.accept(this);
+
+        rt.getObjectsView().TXBegin();
+        try {
+            addrs.add(startCheckpoint());
+            addrs.addAll(appendObjectState());
+            addrs.add(finishCheckpoint());
+            return addrs;
+        } finally {
+            rt.getObjectsView().TXAbort();
+        }
+    }
+
+    /** Append a checkpoint START record to this object's stream.
+     *
+     *  Corfu client transaction management, if desired, is the
+     *  caller's responsibility.
+     *
+     * @return Global log address of the START record.
+     */
+    public long startCheckpoint() {
+        startTime = LocalDateTime.now();
+        AbstractTransactionalContext context = TransactionalContext.getCurrentContext();
+        long txBeginGlobalAddress = context.getSnapshotTimestamp();
+
+        this.mdKV.put(CheckpointEntry.CheckpointDictKey.START_TIME, startTime.toString());
+        this.mdKV.put(CheckpointEntry.CheckpointDictKey.START_LOG_ADDRESS, Long.toString(txBeginGlobalAddress));
+
+        ImmutableMap<CheckpointEntry.CheckpointDictKey,String> mdKV = ImmutableMap.copyOf(this.mdKV);
```

Comment:
1. Will do.

2. I was seeing racy buggy results far down the road, if the dictionary was mutable.